preshed/bloom.pyx

```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    cdef uint32_t safe_seed = int.from_bytes(seed.to_bytes(8, 'big')[-4:], 'big')
+    bloom.seed = safe_seed
+
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
+        buflen += 1
+    contents = struct.unpack(f"<{buflen}Q", data[40:])
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+    for i in range(buflen):
+        bloom.bitfield[i] = contents[i]
+
+
+cdef void bloom_from_bytes_legacy(Pool mem, BloomStruct* bloom, bytes data):
+    # Older versions of this library used the array module with type L for
+    # serialization. Types in array guarantee a minimum size, not an actual
+    # size, and it turns out L is 8 bytes on Linux and most platforms, but 4 on
+    # Windows. 
+
+    # As a separate issue, due to bits/bytes confusion, each container in the
+    # serialized data has only one byte actually used.
+
+    # The code in this function reads in data in the old format and converts it
+    # to the current format losslessly. It also packs the significant bytes into
+    # contiguous memory.
+
+    # on non-Windows platforms
+    unit = "Q"
+    unit_size = 8 # size of container in bytes
+    offset = unit_size * 3
+    hcount, length, seed = struct.unpack("<QQQ", data[0:offset])
+
+    decode_len = length // unit_size # number of units to unpack
+
+    if length != len(data) - 24:
+        # This can happen if the data was serialized on Windows, where the units
+        # were 32bit rather than 64bit.
+        unit = "L"
+        unit_size = 4
+        offset = unit_size * 3
+        hcount, length, seed = struct.unpack("<LLL", data[0:offset])
+
+        # The length was the number bytes in memory. But because of the
+        # platform size issue, the actual serialized bytes is half that.
+        assert length // 2 == len(data) - offset, "Length is invalid"
+
+        decode_len = length // (2 * unit_size)
+
+    bloom.hcount = hcount
+    bloom.length = length
+    cdef uint32_t safe_seed = int.from_bytes(seed.to_bytes(8, 'big')[-4:], 'big')
+    bloom.seed = safe_seed
+
+    # This is tricky - to remove empty space we're going to map bytes into
+    # containers. On Windows or Linux, length is both the number of significant
+    # bits in the bitfield and the number of bytes when the bitfield was in
+    # memory in the old format. In our output, length will be the bitfield
+    # length in bits.
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
+        buflen += 1
+    contents = struct.unpack(f"<{decode_len}{unit}", data[offset:])
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+
+    # Each item in contents provides one significant byte, so we'll copy it
+    # into the containers.
+    for i in range(len(contents)):
+        block = i // sizeof(key_t)
+        idx = i % sizeof(key_t)
+        bloom.bitfield[block] |= contents[i] << (8 * idx)
```

Comment:
Could you explain what is going on here? Aren't the lengths of `bloom.bitfield` and `contents` the same (`buf_len`)? Why is the index divided by the key size for the block and the shifted in bytes by the modulus?

Reply:
This is removing the gaps from the input data to make it contiguous by repacking the data.

If the number of bits in the input is a multiple of `KEY_BITS`, then `bloom.bitfield` will be 1/8 the size of `contents` in memory. This would be awful if all of the bits in `contents` were used, but because of the bug, each item in `contents` is only using one byte, and the others are all 0 / never written to.

Pulling from the tests, it works a bit like this... (this is similar to the explanation above)

```
AA000000 BB000000 CC000000 DD000000
... converted to ...
AABBCCDD
```

In this case, `block` is which container index to go to, and `idx` is the offset within the container.